src/GameServer/Systems/Inventory/InventoryManager.cs

```diff
@@ -55,7 +56,143 @@ private int GetVirtualItemValue(int itemId)
+            }
+            int moraCost = expGain / 10;
+            expGain += expGainFree;
+            int exp = weapon.Exp;
+            int level = weapon.Level;
+            int oldLevel = level;
+            if (Owner.PlayerProperties.GetValueOrDefault(PlayerProperty.PROP_PLAYER_SCOIN) < moraCost)
+                return;
+            await Owner.PropManager.PayMoraAsync(moraCost); 
+            if (await RemoveItemsByParamData(xpMats))
+            {
+                int reqExp = GameData.WeaponLevelDataMap[level].requiredExps[weapon.ItemData.rankLevel-1];               
+                while (expGain > 0 && level < weapon.promoteData.unlockMaxLevel)
+                {
+                    int toGain = Math.Min(expGain, reqExp - exp);
+                    exp += toGain;
+                    weapon.setTotalExp(weapon.TotalExp + toGain);
+                    expGain -= toGain;
+                    if (exp >= reqExp)
+                    {
+                        exp = 0;
+                        level += 1;
+                        reqExp = GameData.WeaponLevelDataMap[level].requiredExps[weapon.ItemData.rankLevel-1];
+                    }
+                
+                }
+                weapon.setLevel(level);
+                weapon.setExp(exp);
+                await (SubInventories[ItemType.ITEM_WEAPON] as WeaponTab).updateWeaponAsync(weapon);
+                await Owner.SendPacketAsync(new PacketStoreItemChangeNotify(weapon));
+                await Owner.SendPacketAsync(new PacketWeaponUpgradeRsp(weapon, oldLevel, leftoverOres));
+            }
+        }
 
+        public async Task<bool> promoteWeaponAsync(ulong targetWeaponGuid)
+        {
+            WeaponItem weapon = GuidMap[targetWeaponGuid] as WeaponItem;
+            GameData.WeaponPromoteDataMap.TryGetValue(Tuple.Create(weapon.ItemData.weaponPromoteId, weapon.PromoteLevel + 1), out WeaponPromoteData? promoteData);
+            if (weapon is null || promoteData is null || Owner.PlayerProperties[PlayerProperty.PROP_PLAYER_LEVEL] < promoteData.requiredPlayerLevel)
+                return false;
+            if (Owner.PlayerProperties.GetValueOrDefault(PlayerProperty.PROP_PLAYER_SCOIN) < promoteData.coinCost)
+                return false;
+            await Owner.PropManager.PayMoraAsync(promoteData.coinCost);
+            if(await PayPromoteCostAsync(promoteData.costItems, ActionReason.WeaponPromote))
+            {
+                int oldPromote = weapon.PromoteLevel;
+                weapon.Promote();
+                await (SubInventories[ItemType.ITEM_WEAPON] as WeaponTab).updateWeaponAsync(weapon);
+                await Owner.SendPacketAsync(new PacketStoreItemChangeNotify(weapon));
+                await Owner.SendPacketAsync(new PacketWeaponPromoteRsp(weapon, oldPromote));
+                return true;
+            }
+            return false;
+        }
+        public async Task<bool> RemoveItemsByParamData(List<ItemParamData> itemDataList)
```

Comment:
This function will still remove some items even if it ultimately returns false.
```suggestion
        public async Task<bool> RemoveItemsByParamData(IEnumerable<ItemParamData> itemDataList)
```